tesseract_common/src/utils.cpp

```diff
@@ -0,0 +1,286 @@
+    CONSOLE_BRIDGE_logError("Invalid %s attribute '%s'", xml_element->Name(), name);
+  }
+  else if (status == tinyxml2::XML_NO_ATTRIBUTE)
+  {
+    CONSOLE_BRIDGE_logError("Missing %s required attribute '%s'", xml_element->Name(), name);
+  }
+  else if (status == tinyxml2::XML_WRONG_ATTRIBUTE_TYPE)
+  {
+    CONSOLE_BRIDGE_logError("Invalid %s attribute type '%s'", xml_element->Name(), name);
+  }
+
+  return status;
+}
+
+tinyxml2::XMLError QueryDoubleAttributeRequired(const tinyxml2::XMLElement* xml_element,
+                                                const char* name,
+                                                double& value)
+{
+  tinyxml2::XMLError status = xml_element->QueryDoubleAttribute(name, &value);
+
+  if (status != tinyxml2::XML_NO_ATTRIBUTE && status != tinyxml2::XML_SUCCESS)
+  {
+    CONSOLE_BRIDGE_logError("Invalid %s attribute '%s'", xml_element->Name(), name);
+  }
+  else if (status == tinyxml2::XML_NO_ATTRIBUTE)
+  {
+    CONSOLE_BRIDGE_logError("Missing %s required attribute '%s'", xml_element->Name(), name);
+  }
+  else if (status == tinyxml2::XML_WRONG_ATTRIBUTE_TYPE)
+  {
+    CONSOLE_BRIDGE_logError("Invalid %s attribute type '%s'", xml_element->Name(), name);
+  }
+
+  return status;
+}
+
+tinyxml2::XMLError QueryIntAttributeRequired(const tinyxml2::XMLElement* xml_element, const char* name, int& value)
+{
+  tinyxml2::XMLError status = xml_element->QueryIntAttribute(name, &value);
+
+  if (status != tinyxml2::XML_NO_ATTRIBUTE && status != tinyxml2::XML_SUCCESS)
+  {
+    CONSOLE_BRIDGE_logError("Invalid %s attribute '%s'", xml_element->Name(), name);
+  }
+  else if (status == tinyxml2::XML_NO_ATTRIBUTE)
+  {
+    CONSOLE_BRIDGE_logError("Missing %s required attribute '%s'", xml_element->Name(), name);
+  }
+  else if (status == tinyxml2::XML_WRONG_ATTRIBUTE_TYPE)
+  {
+    CONSOLE_BRIDGE_logError("Invalid %s attribute type '%s'", xml_element->Name(), name);
+  }
+
+  return status;
+}
+
+bool almostEqualRelativeAndAbs(double a, double b, double max_diff, double max_rel_diff)
+{
+  double diff = std::fabs(a - b);
+  if (diff <= max_diff)
+    return true;
+
+  a = std::fabs(a);
+  b = std::fabs(b);
+  double largest = (b > a) ? b : a;
+
+  return (diff <= largest * max_rel_diff);
+}
+
+bool almostEqualRelativeAndAbs(const Eigen::Ref<const Eigen::VectorXd>& v1,
+                               const Eigen::Ref<const Eigen::VectorXd>& v2,
+                               double max_diff,
+                               double max_rel_diff)
+{
+  if (v1.size() == 0 && v2.size() == 0)
+    return true;
+  if (v1.size() != v2.size())
+    return false;
+
+  Eigen::ArrayWrapper<const Eigen::Ref<const Eigen::VectorXd>> a1 = v1.array();
+  Eigen::ArrayWrapper<const Eigen::Ref<const Eigen::VectorXd>> a2 = v2.array();
+
+  auto diff_abs = (a1 - a2).abs();
+  double diff = diff_abs.maxCoeff();
+  if (diff <= max_diff)
+    return true;
+
+  return (diff_abs <= (max_rel_diff * a1.abs().max(a2.abs()))).all();
+}
+
```

Comment:
Explicit instantiation of isNumeric for a couple of types?

Reply:
I will add this to the PR #592

Reply:
done.